client/history/client.go

```diff
@@ -846,24 +856,37 @@ func (c *clientImpl) GetReplicationMessages(
 				}
 				return
 			}
-			respChan <- resp
-		}(requestContext, client, req)
+			_, responseInfo := rpc.ContextWithResponseInfo(requestContext)
+			respChan <- &getReplicationMessagesWithSize{
+				response: resp,
+				size:     responseInfo.Size,
+			}
+		}(ctx, client, req)
 	}
 
 	wg.Wait()
 	close(respChan)
 	close(errChan)
 
+	var err error
+	if len(errChan) > 0 {
+		err = <-errChan
```

Comment:
we can return early here, and skip the logic for preparing the result for the rest of the shards?